File: app/term.py
"""Real PTY attached to `tmux attach-session`, bridged to a WebSocket.

Wire protocol
- client → server: **binary** frames are raw keystrokes; **text** frames are JSON control
  messages (``{"type":"resize","cols":N,"rows":N}``).
- server → client: **binary** frames are raw PTY output; **text** frames are JSON status
  (``{"type":"ready"}`` / ``{"type":"exit","reason":"…"}``).
"""

from __future__ import annotations

import asyncio
import contextlib
import fcntl
import subprocess
import json
import os
import pty
import select
import signal
import struct
import termios
import threading
import time

from fastapi import APIRouter, WebSocket

from . import tmux
from .config import home

READ_BUF = 8192
# Bounded so a client that stops reading applies backpressure to the PTY instead of
# letting a runaway `cat` of a huge file balloon our memory.
OUTPUT_QUEUE = 512
# How long a read waits for company before being sent. Long enough to gather a burst,
# short enough that a keystroke echo still feels immediate.
COALESCE = 0.008
COALESCE_MAX = 256 * 1024

router = APIRouter()


def clamp(n: int) -> int:
    return max(2, min(int(n), 1000))


def set_winsize(fd: int, rows: int, cols: int) -> None:
    fcntl.ioctl(fd, termios.TIOCSWINSZ, struct.pack("HHHH", rows, cols, 0, 0))


def others_attached(sock: tmux.Socket, session: str) -> int:
    """How many clients are already on this session."""
    try:
        p = subprocess.run(
            ["tmux", *sock.args(), "list-clients", "-t", f"={session}", "-F", "#{client_name}"],
            capture_output=True, text=True, timeout=4,
        )
    except (OSError, subprocess.SubprocessError):
        return 0
    return len([ln for ln in p.stdout.splitlines() if ln.strip()]) if p.returncode == 0 else 0


def window_size(sock: tmux.Socket, session: str) -> tuple[int, int] | None:
    """The grid tmux is actually drawing for this session."""
    try:
        p = subprocess.run(
            # No `=` here: display-message resolves a pane target and returns nothing
            # for the exact-match form on this tmux. The name came from list-sessions,
            # so it is already exact.
            ["tmux", *sock.args(), "display-message", "-p", "-t", session,
             "#{window_width}x#{window_height}"],
            capture_output=True, text=True, timeout=4,
        )
    except (OSError, subprocess.SubprocessError):
        return None
    try:
        w, h = p.stdout.strip().split("x")
        return int(w), int(h)
    except ValueError:
        return None


def attach_argv(session: str, flags: list[str], sock: tmux.Socket) -> list[str]:
    # Socket selection is a global flag: it has to precede the command, and it decides
    # which server we attach to at all. `-u` forces UTF-8 regardless of the server's
    # locale. The session name is one argv element — never a shell string.
    return ["tmux", *sock.args(), "-u", "attach-session", *flags, "-t", session]


def child_env() -> dict[str, str]:
    """Build the environment explicitly rather than inheriting wholesale.

    TMUX/TMUX_PANE is the one that matters: argus is very likely started from
    inside a tmux pane, and tmux refuses to attach when it thinks it would nest.
    """
    env = {k: v for k, v in os.environ.items() if k not in ("TMUX", "TMUX_PANE", "TERM")}
    env["TERM"] = "xterm-256color"
    return env


def spawn(argv: list[str], env: dict[str, str], rows: int, cols: int) -> tuple[int, int]:
    """Fork a child on a fresh PTY and exec into it. Returns (pid, master fd).

    The size is set on the slave *before* the fork, so tmux sees the phone's geometry
    from its very first draw instead of resizing a moment later.
    """
    master, slave = pty.openpty()
    set_winsize(slave, rows, cols)
    pid = os.fork()
    if pid == 0:  # child
        try:
            os.close(master)
            # setsid + TIOCSCTTY + dup2 onto 0/1/2. Without a controlling terminal tmux
            # refuses to attach.
            os.login_tty(slave)
            os.chdir(str(home()))
            os.execvpe(argv[0], argv, env)
        except BaseException:
            os._exit(127)
    os.close(slave)
    return pid, master


def _reader(fd: int, loop: asyncio.AbstractEventLoop, queue: asyncio.Queue) -> None:
    """Blocking reads on their own thread, handing chunks to the event loop.

    Reads are coalesced for a few milliseconds before being handed over. A program
    printing steadily produces dozens of small reads a second, and one WebSocket frame
    each means the browser wakes, decodes and repaints dozens of times to draw what is
    really one update — which is what reads as flicker over a slow link. Awaiting the
    queue put is what propagates backpressure to the PTY.
    """
    while True:
        try:
            data = os.read(fd, READ_BUF)
        except OSError:
            data = b""  # EIO: the child is gone and the pty has no slave left
        if not data:
            break

        # Whatever else arrives within the window joins this frame, up to a cap so a
        # flood still gets delivered in pieces the browser can render.
        deadline = time.monotonic() + COALESCE
        while len(data) < COALESCE_MAX:
            left = deadline - time.monotonic()
            if left <= 0 or not select.select([fd], [], [], left)[0]:
                break
            try:
                more = os.read(fd, READ_BUF)
            except OSError:
                break
            if not more:
                break
            data += more

        try:
            asyncio.run_coroutine_threadsafe(queue.put(data), loop).result()
        except Exception:
            return
    with contextlib.suppress(Exception):
        asyncio.run_coroutine_threadsafe(queue.put(None), loop).result()


def _terminate(pid: int, master: int) -> None:
    """Detach our client and let go of the PTY.

    Killing the client is enough — the tmux *session* and everything running in it must
    survive. That is the entire point of the product.
    """
    with contextlib.suppress(ProcessLookupError, OSError):
        os.kill(pid, signal.SIGHUP)
    for _ in range(50):
        try:
            if os.waitpid(pid, os.WNOHANG)[0] == pid:
                break
        except ChildProcessError:
            break
        threading.Event().wait(0.01)
    else:
        with contextlib.suppress(ProcessLookupError, OSError):
            os.kill(pid, signal.SIGKILL)
        with contextlib.suppress(ChildProcessError, OSError):
            os.waitpid(pid, 0)
    with contextlib.suppress(OSError):
        os.close(master)


@router.websocket("/ws/tmux/{session}")
async def terminal(websocket: WebSocket, session: str, cols: int = 80, rows: int = 24) -> None:
    state = websocket.app.state
    await websocket.accept()

    # Only names tmux itself reported are acceptable. The PTY spawn takes an argv (no
    # shell), so this is not about injection — it is about not handing tmux a target we
    # never listed.
    if not await asyncio.to_thread(tmux.session_exists, state.socket, session):
        await websocket.send_text(
            json.dumps({"type": "exit", "reason": f"no tmux session named {session!r}"})
        )
        return await websocket.close()

    cols, rows = clamp(cols), clamp(rows)

    # `auto`: take the window size when nobody else is here, leave it alone when someone
    # is. tmux sizes a window for all its clients at once, so two devices on one session
    # otherwise resize it back and forth at each other — which is the flicker, and the
    # feeling that the two are never quite showing the same thing.
    flags = state.cfg.attach_flags()
    if state.cfg.resize_policy == "auto":
        crowded = await asyncio.to_thread(others_attached, state.socket, session)
        flags = ["-f", "ignore-size"] if crowded else []

    pid, master = spawn(
        attach_argv(session, flags, state.socket), child_env(), rows, cols
    )

    loop = asyncio.get_running_loop()
    queue: asyncio.Queue = asyncio.Queue(OUTPUT_QUEUE)
    thread = threading.Thread(target=_reader, args=(master, loop, queue), daemon=True)
    thread.start()

    # Tell the client whether it was allowed to set the size. When it was not, it needs
    # the grid tmux is drawing so it can fit that grid by shrinking the font rather than
    # by asking for fewer columns — which is how two devices end up showing the same
    # screen without resizing it at each other.
    fixed = "ignore-size" in flags
    grid = await asyncio.to_thread(window_size, state.socket, session) if fixed else None
    await websocket.send_text(json.dumps({
        "type": "ready",
        "fixed": fixed,
        **({"cols": grid[0], "rows": grid[1]} if grid else {}),
    }))

    async def pump_out() -> str:
        while True:
            chunk = await queue.get()
            # tmux exited or the PTY closed — tell the client instead of leaving it
            # staring at a frozen terminal.
            if chunk is None:
                return "tmux exited"
            await websocket.send_bytes(chunk)

    async def pump_in() -> str:
        while True:
            message = await websocket.receive()
            if message["type"] == "websocket.disconnect":
                return "client closed"
            data = message.get("bytes")
            if data is not None:
                await asyncio.to_thread(os.write, master, data)
                continue
            text = message.get("text")
            if text is None:
                continue
            control = None
            with contextlib.suppress(ValueError):
                control = json.loads(text)
            if isinstance(control, dict) and control.get("type") == "resize":
                with contextlib.suppress(OSError, KeyError, TypeError, ValueError):
                    set_winsize(master, clamp(control["rows"]), clamp(control["cols"]))
            else:
                # A browser that sends keystrokes as text still works.
                await asyncio.to_thread(os.write, master, text.encode("utf-8"))

    tasks = [asyncio.create_task(pump_out()), asyncio.create_task(pump_in())]
    try:
        done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
        for task in pending:
            task.cancel()
        finished = done.pop()
        try:
            reason = finished.result()
        except Exception:
            reason = "connection error"
    finally:
        _terminate(pid, master)
        await asyncio.to_thread(thread.join, 2.0)

    with contextlib.suppress(Exception):
        await websocket.send_text(json.dumps({"type": "exit", "reason": reason}))
        await websocket.close()

File: legacy-rust/src/ws/term.rs
//! Real PTY attached to `tmux attach-session`, bridged to a WebSocket.
//!
//! Wire protocol
//! - client → server: **binary** frames are raw keystrokes; **text** frames are JSON control
//!   messages (`{"type":"resize","cols":N,"rows":N}`).
//! - server → client: **binary** frames are raw PTY output; **text** frames are JSON status
//!   (`{"type":"ready"}` / `{"type":"exit","reason":"…"}`).

use axum::{
    extract::{
        ws::{Message, WebSocket, WebSocketUpgrade},
        Path, Query, State,
    },
    http::StatusCode,
    response::{IntoResponse, Response},
};
use futures_util::{SinkExt, StreamExt};
use portable_pty::{CommandBuilder, MasterPty, PtySize};
use serde::Deserialize;
use std::io::{Read, Write};
use std::sync::{Arc, Mutex};

use crate::{api::tmux, config::ResizePolicy, AppState};

const READ_BUF: usize = 8192;
/// Bounded so a client that stops reading applies backpressure to the PTY instead of
/// letting a runaway `cat` of a huge file balloon our memory.
const OUTPUT_QUEUE: usize = 512;

#[derive(Deserialize)]
pub struct TermQuery {
    #[serde(default = "default_cols")]
    pub cols: u16,
    #[serde(default = "default_rows")]
    pub rows: u16,
}

fn default_cols() -> u16 {
    80
}
fn default_rows() -> u16 {
    24
}

#[derive(Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
enum ClientControl {
    Resize { cols: u16, rows: u16 },
}

pub async fn handler(
    ws: WebSocketUpgrade,
    Path(session): Path<String>,
    Query(q): Query<TermQuery>,
    State(st): State<AppState>,
) -> Response {
    // Only names tmux itself reported are acceptable. The PTY spawn takes an argv (no
    // shell), so this is not about injection — it is about not handing tmux a target we
    // never listed, and about returning a clean 404 before upgrading the connection.
    let name = session.clone();
    let sock = st.socket.clone();
    let probe = sock.clone();
    let exists = tokio::task::spawn_blocking(move || tmux::session_exists(&probe, &name))
        .await
        .unwrap_or(false);
    if !exists {
        return (StatusCode::NOT_FOUND, format!("no tmux session named {session:?}")).into_response();
    }

    let policy = st.cfg.resize_policy;
    let cols = q.cols.clamp(2, 1000);
    let rows = q.rows.clamp(2, 1000);

    ws.on_upgrade(move |socket| async move {
        if let Err(e) = bridge(socket, session, cols, rows, policy, sock).await {
            eprintln!("terminal session ended with an error: {e:#}");
        }
    })
}

async fn bridge(
    socket: WebSocket,
    session: String,
    cols: u16,
    rows: u16,
    policy: ResizePolicy,
    sock: tmux::Socket,
) -> anyhow::Result<()> {
    let pty = portable_pty::native_pty_system().openpty(PtySize {
        rows,
        cols,
        pixel_width: 0,
        pixel_height: 0,
    })?;

    let mut child = pty.slave.spawn_command(attach_command(&session, policy, &sock))?;
    // The slave fd must go, otherwise the master never sees EOF when tmux exits.
    drop(pty.slave);

    let master: Arc<Mutex<Box<dyn MasterPty + Send>>> = Arc::new(Mutex::new(pty.master));
    let mut reader = master.lock().unwrap().try_clone_reader()?;
    let mut writer = master.lock().unwrap().take_writer()?;

    // portable-pty gives us blocking handles, so each direction gets its own thread and
    // talks to the async side through a channel.
    let (out_tx, mut out_rx) = tokio::sync::mpsc::channel::<Vec<u8>>(OUTPUT_QUEUE);
    std::thread::spawn(move || {
        let mut buf = [0u8; READ_BUF];
        loop {
            match reader.read(&mut buf) {
                Ok(0) | Err(_) => break,
                Ok(n) => {
                    if out_tx.blocking_send(buf[..n].to_vec()).is_err() {
                        break;
                    }
                }
            }
        }
    });

    let (in_tx, in_rx) = std::sync::mpsc::channel::<Vec<u8>>();
    std::thread::spawn(move || {
        while let Ok(data) = in_rx.recv() {
            if writer.write_all(&data).is_err() || writer.flush().is_err() {
                break;
            }
        }
    });

    let (mut ws_tx, mut ws_rx) = socket.split();
    let _ = ws_tx.send(Message::Text(r#"{"type":"ready"}"#.into())).await;

    let reason = loop {
        tokio::select! {
            chunk = out_rx.recv() => match chunk {
                Some(data) => {
                    if ws_tx.send(Message::Binary(data.into())).await.is_err() {
                        break "client went away";
                    }
                }
                // tmux exited or the PTY closed — tell the client instead of leaving it
                // staring at a frozen terminal.
                None => break "tmux exited",
            },
            msg = ws_rx.next() => match msg {
                Some(Ok(Message::Binary(b))) => {
                    if in_tx.send(b.to_vec()).is_err() {
                        break "pty closed";
                    }
                }
                // A browser that sends keystrokes as text still works.
                Some(Ok(Message::Text(t))) => match serde_json::from_str::<ClientControl>(&t) {
                    Ok(ClientControl::Resize { cols, rows }) => {
                        let size = PtySize {
                            rows: rows.clamp(2, 1000),
                            cols: cols.clamp(2, 1000),
                            pixel_width: 0,
                            pixel_height: 0,
                        };
                        if let Ok(m) = master.lock() {
                            let _ = m.resize(size);
                        }
                    }
                    Err(_) => {
                        if in_tx.send(t.as_bytes().to_vec()).is_err() {
                            break "pty closed";
                        }
                    }
                },
                Some(Ok(Message::Close(_))) | None => break "client closed",
                Some(Err(_)) => break "connection error",
                Some(Ok(_)) => {} // ping/pong handled by axum
            },
        }
    };

    let _ = ws_tx
        .send(Message::Text(
            serde_json::json!({ "type": "exit", "reason": reason }).to_string().into(),
        ))
        .await;
    let _ = ws_tx.send(Message::Close(None)).await;

    // Detaching the client is enough — the tmux *session* and everything running in it
    // must survive, that is the entire point of the product.
    let _ = child.kill();
    let _ = child.wait();
    Ok(())
}

fn attach_command(session: &str, policy: ResizePolicy, sock: &tmux::Socket) -> CommandBuilder {
    let mut cmd = CommandBuilder::new("tmux");
    // Socket selection is a global flag: it has to precede the command, and it decides
    // which server we attach to at all.
    for a in sock.args() {
        cmd.arg(a);
    }
    cmd.arg("-u"); // force UTF-8 regardless of the server's locale
    cmd.arg("attach-session");
    for f in policy.attach_flags() {
        cmd.arg(f);
    }
    cmd.arg("-t");
    cmd.arg(session);

    // Build the environment explicitly rather than inheriting: `env_clear` first, then
    // copy across what we want. Overriding alone is not enough — an inherited value we
    // never mention would still reach the child.
    //
    // TMUX/TMUX_PANE is the one that matters: tmux-companion is very likely started from
    // inside a tmux pane, and tmux refuses to attach when it thinks it would nest.
    cmd.env_clear();
    for (k, v) in std::env::vars() {
        if k == "TMUX" || k == "TMUX_PANE" || k == "TERM" {
            continue;
        }
        cmd.env(k, v);
    }
    cmd.env("TERM", "xterm-256color");
    cmd.cwd(crate::config::home());
    cmd
}

#[cfg(test)]
mod tests {
    use super::*;

    fn argv(cmd: &CommandBuilder) -> Vec<String> {
        cmd.get_argv().iter().map(|a| a.to_string_lossy().into_owned()).collect()
    }

    fn env_of(cmd: &CommandBuilder, key: &str) -> Option<String> {
        cmd.iter_full_env_as_str()
            .find(|(k, _)| *k == key)
            .map(|(_, v)| v.to_string())
    }

    #[test]
    fn adapt_attaches_plainly() {
        let c = attach_command("claude", ResizePolicy::Adapt, &tmux::Socket::default());
        assert_eq!(argv(&c), vec!["tmux", "-u", "attach-session", "-t", "claude"]);
    }

    #[test]
    fn preserve_adds_ignore_size_so_other_clients_keep_their_geometry() {
        let c = attach_command("claude", ResizePolicy::Preserve, &tmux::Socket::default());
        assert_eq!(
            argv(&c),
            vec!["tmux", "-u", "attach-session", "-f", "ignore-size", "-t", "claude"]
        );
    }

    #[test]
    fn session_name_is_passed_as_a_single_argument_never_a_shell_string() {
        let c = attach_command("weird; rm -rf /", ResizePolicy::Adapt, &tmux::Socket::default());
        let a = argv(&c);
        assert_eq!(a.last().unwrap(), "weird; rm -rf /");
        assert_eq!(a.len(), 5);
    }

    #[test]
    fn a_configured_socket_is_what_we_attach_to() {
        let c = attach_command("claude", ResizePolicy::Adapt, &tmux::Socket::new(Some("tmuxc-test")));
        assert_eq!(
            argv(&c),
            vec!["tmux", "-L", "tmuxc-test", "-u", "attach-session", "-t", "claude"],
            "-L must precede the command, or tmux drives the default server instead"
        );
    }

    #[test]
    fn tmux_env_is_stripped_so_attaching_from_inside_tmux_works() {
        std::env::set_var("TMUX", "/tmp/tmux-1000/default,123,4");
        std::env::set_var("TMUX_PANE", "%7");
        let c = attach_command("claude", ResizePolicy::Adapt, &tmux::Socket::default());
        let leaked = env_of(&c, "TMUX").or_else(|| env_of(&c, "TMUX_PANE"));
        std::env::remove_var("TMUX");
        std::env::remove_var("TMUX_PANE");
        assert_eq!(leaked, None, "TMUX must not reach the child, or tmux refuses to nest");
    }

    #[test]
    fn term_is_forced_to_a_256_colour_value() {
        std::env::set_var("TERM", "dumb");
        let c = attach_command("claude", ResizePolicy::Adapt, &tmux::Socket::default());
        assert_eq!(env_of(&c, "TERM").as_deref(), Some("xterm-256color"));
    }

    #[test]
    fn ordinary_environment_still_reaches_the_child() {
        std::env::set_var("TMUXC_PROBE", "kept");
        let c = attach_command("claude", ResizePolicy::Adapt, &tmux::Socket::default());
        let got = env_of(&c, "TMUXC_PROBE");
        std::env::remove_var("TMUXC_PROBE");
        assert_eq!(got.as_deref(), Some("kept"), "clearing the env must not strip PATH/HOME/LANG");
    }

    #[test]
    fn resize_control_messages_parse() {
        let m: ClientControl = serde_json::from_str(r#"{"type":"resize","cols":100,"rows":30}"#).unwrap();
        let ClientControl::Resize { cols, rows } = m;
        assert_eq!((cols, rows), (100, 30));
        assert!(serde_json::from_str::<ClientControl>(r#"{"type":"nope"}"#).is_err());
        assert!(serde_json::from_str::<ClientControl>("ls -la\n").is_err());
    }
}
